### PR TITLE
Support for GitHub apps user installation convergence

### PR DESCRIPTION
**Pull request by kipz**

### CODE REVIEW

File: lib/converge/api.ts [hunk 1]
```diff
@@ -108,3 +111,19 @@ export async function loadProvider(graphClient: GraphClient,
...
+    installations: GitHubAppInstallationInput[]):
+    Promise<SaveGitHubAppUserInstallations.SaveGitHubAppUserInstallations> {
+        return (await graphClient.mutate<SaveGitHubAppUserInstallations.Mutation, SaveGitHubAppUserInstallations.Variables>({
+            name: "SaveGitHubAppUserInstallations",
+            variables: {
+                installations,
+                pid: scmId.provider.id,
+                userId: scmId.id,
+            },
+            options: QueryNoCacheOptions,
```
Comment (cdupuis) [inline lib/converge/api.ts]: That's a no-op on mutations. Please use `MutationNoCacheOptions` or drop it.

File: lib/converge/api.ts [hunk 2]
```diff
@@ -108,3 +111,19 @@ export async function loadProvider(graphClient: GraphClient,
         options: QueryNoCacheOptions,
     })).SCMProvider[0];
 }
+
+export async function saveGitHubAppUserInstallations(
```
Comment (cdupuis) [inline lib/converge/api.ts]: Please align parameters with function ():

File: lib/graphql/subscription/gitHubAppScmId.graphql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+subscription OnGibHubAppScmId {
+  SCMId {
+    id
+    login @required
+    credential {
+      secret @required
+    }
+    provider {
+        apiUrl
+        providerType @matches(re: "github_com_app")
```
Comment (cdupuis) [inline lib/graphql/subscription/gitHubAppScmId.graphql]: Why do we need the non-standard `matches` directive here and can't use `provider(providerType = github_com_app)`?

File: lib/converge/convergeGitHub.ts [hunk 4]
```diff
@@ -123,3 +127,20 @@ function onScmProvider(options: ConvergenceOptions): EventHandlerRegistration<On
...
     };
 }
+
+/**
+ * EventHandlerRegistration listening for new SCMId events for github apps, and triggering user org convergence
+ */
+function onGitHubAppsScmId(options: ConvergenceOptions): EventHandlerRegistration<OnGibHubAppScmId.Subscription> {
+    return {
+        name: "ConvergeGitHubAppsOnScmId",
+        subscription: GraphQL.subscription({
```
Comment (cdupuis) [inline lib/converge/convergeGitHub.ts]: Use `GrapQL.subscription("OnGibHubAppScmId")` as short-hand here.

**Review by cdupuis — CHANGES_REQUESTED** [review]
Couple of changes.

File: lib/graphql/subscription/gitHubAppScmId.graphql (hunk 3 above)
Comment (kipz) [reply lib/graphql/subscription/gitHubAppScmId.graphql]: @cdupuis I didn't think `@matches` was non-standard, but also, the provider sub-query doesn't take provider type as an argument. If it did, I could use `@required` - is that considered more standard?

File: lib/converge/api.ts (hunk 2 above)
Comment (kipz) [reply lib/converge/api.ts]: @cdupuis done - though I would expect a linter/autofix to catch such things.

File: lib/graphql/subscription/gitHubAppScmId.graphql (hunk 3 above)
Comment (cdupuis) [reply lib/graphql/subscription/gitHubAppScmId.graphql]: I'd consider the argument more standard considering all the other subscriptions we use throughout. The fact providerType is not an argument is confusing IMHO.

Comment (kipz) [reply lib/graphql/subscription/gitHubAppScmId.graphql]: The graphql API was always like that I think - I'll raise an issue to get that changed.

Can we go ahead as is until that's been sorted?

**Comment by cdupuis** [comment]
Pull request rebase to `b255f4b` by @cdupuis failed because of following conflicting files:
- `package-lock.json`
- `package.json`

**Comment by cdupuis** [comment]
Pull request rebase to `c8e76e0` by @cdupuis failed because of following conflicting files:
- `package-lock.json`
- `package.json`

**Review by cdupuis — APPROVED** [review]

**Comment by cdupuis** [comment]
Pull request auto merged by Atomist.

* 1 approved review by @cdupuis
* 2 successful checks

[atomist:generated] [auto-merge:on-approve]

**Pull request merged**